### PR TITLE
refactor health check into HealthScope controller

### LINKED ISSUE #914: [Feature] health check logic should refactor from app-controller into healthscope controller
Now we have implemented health check policy by using cue in https://github.com/oam-dev/kubevela/pull/861.

But the code is coupled in Application controller, the main logic is  a periodically check for health. But the Application Controller should not have this dependency, it should not add a cron-logic for health check, this should be done in health scope controller.

### PR DESCRIPTION
**Pull request by captainroy-hy**
Signed-off-by: Yue Wang <seiwy2010@gmail.com>

**What this PR does / why we need it**:

move health check logic from app controller to health scope controller

**Which issue(s) this PR fixes**:

Fixes #914 

**Special notes for your reviewer**:

- [x] add CUE-based health check in HealthScope controller
- [x] test

### CODE REVIEW

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go [hunk 1]
```diff
@@ -116,15 +116,49 @@ func (a *AppManifestsDispatcher) Dispatch(ctx context.Context, manifests []*unst
...
+			return err
+		}
+		copyS := scopeObj.DeepCopy()
+
+		tmp := map[string]interface{}{}
+		for k, v := range compRefs {
+			tmp[k] = v
+		}
+		// TODO hard-code reference field path should be replcaced with getting from scope definition
+		if err := unstructured.SetNestedMap(scopeObj.Object, tmp, "spec", "appReferences", appName); err != nil {
```
Comment (captainroy-hy) [inline pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go:136]: @wonderflow call for help
As shown here, I hope to set ComponentReferences to an unsturctued object's specific field(spec.appReferences), but it failed for "Cannot deep copy ComponentReferences".

File: apis/core.oam.dev/v1alpha2/core_scope_types.go [hunk 2]
```diff
@@ -47,8 +47,24 @@ type HealthScopeSpec struct {
 	// ProbeInterval is the amount of time in seconds between probing tries.
 	ProbeInterval *int32 `json:"probe-interval,omitempty"`
 
+	// AppReferences records references of applications' components
+	AppReferences `json:"appReferences,omitempty"`
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha2/core_scope_types.go:51]:
```suggestion
   AppRef   AppReferences `json:"appReferences,omitempty"`
```


Try this ?

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go [hunk 3]
```diff
@@ -116,15 +116,49 @@ func (a *AppManifestsDispatcher) Dispatch(ctx context.Context, manifests []*unst
...
+			return err
+		}
+	}
 	return nil
 }
 
-// DereferenceScopes remove workload reference from scopes' workloadRefPath
-func (a *AppManifestsDispatcher) DereferenceScopes(ctx context.Context, wlRef *v1.ObjectReference, scopes []*v1.ObjectReference) error {
-	// TODO handle scopes
+// DereferenceScopes remove an application's all workload references from scopes' workload reference field
```
Comment (hongchaodeng) [inline pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go:146]:
```suggestion
// DereferenceScopes removes all of the workload references from an applicatin's scopes workload reference field
```

**Review by hongchaodeng — COMMENTED** [review]
I think the current health check policy could be written in the CUE template of Comp/Trait Def, not necessarily Scope. My question is, how does application controller detect such logic in comp/traits and put them into scopes?

**Comment by captainroy-hy** [comment]
@wonderflow  Below are samples used by this PR.
The sample health scope checks four comonents from two applications.
```yaml
# healthscope.yaml
apiVersion: core.oam.dev/v1alpha2
kind: HealthScope
metadata:
  name: example-health-scope
spec:
  probe-timeout: 5
  probe-interval: 1
---
# app1.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: test-health-app-1
spec:
  components:
    - name: test-comp-1-1
      type: webservice
      properties:
        image: crccheck/hello-world
        port: 9993
      traits:
        - type: ingress
          properties:
            domain: localhost2
            http:
              "/": 8000
      scopes:
        healthscopes.core.oam.dev: example-health-scope
    - name: test-comp-1-2
      type: webservice
      properties:
        image: crccheck/hello-world
        port: 9993
      scopes:
        healthscopes.core.oam.dev: example-health-scope
---
# app2.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: test-health-app-2
spec:
  components:
    - name: test-comp-2-1
      type: webservice
      properties:
        image: crccheck/hello-world
        port: 9993
      traits:
        - type: ingress
          properties:
            domain: localhost2
            http:
              "/": 8000
      scopes:
        healthscopes.core.oam.dev: example-health-scope
    - name: test-comp-2-2
      type: webservice
      properties:
        image: crccheck/hello-world
        port: 9993
      scopes:
        healthscopes.core.oam.dev: example-health-scope

````
And this is example-health-scope after applying
```yaml
apiVersion: core.oam.dev/v1alpha2
kind: HealthScope
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"core.oam.dev/v1alpha2","kind":"HealthScope","metadata":{"annotations":{},"name":"example-health-scope","namespace":"default"},"spec":{"probe-interval":1,"probe-timeout":5}}
  creationTimestamp: "2021-08-03T09:08:27Z"
  generation: 9
  name: example-health-scope
  namespace: default
  resourceVersion: "2582530"
  uid: 0adeba5a-a997-4ef5-b1d3-9e91bdeb8fa7
spec:
  appReferences: # !! Pls help to check this field. It's used to record references of applications' components.
    test-health-app-1:
      test-comp-1-1:
        traits:
        - apiVersion: v1
          kind: Service
          name: test-comp-1-1
        - apiVersion: networking.k8s.io/v1beta1
          kind: Ingress
          name: test-comp-1-1
        workload:
          apiVersion: apps/v1
          kind: Deployment
          name: test-comp-1-1
      test-comp-1-2:
        workload:
          apiVersion: apps/v1
          kind: Deployment
          name: test-comp-1-2
    test-health-app-2:
      test-comp-2-1:
        traits:
        - apiVersion: v1
          kind: Service
          name: test-comp-2-1
        - apiVersion: networking.k8s.io/v1beta1
          kind: Ingress
          name: test-comp-2-1
        workload:
          apiVersion: apps/v1
          kind: Deployment
          name: test-comp-2-1
      test-comp-2-2:
        workload:
          apiVersion: apps/v1
          kind: Deployment
          name: test-comp-2-2
  probe-interval: 1
  probe-timeout: 5
status:
  appHealthConditions:
  - appName: test-health-app-1
    components:
    - componentName: test-comp-1-1
      diagnosis: 'Ready:1/1 '
      healthStatus: HEALTHY
      targetWorkload:
        apiVersion: apps/v1
        kind: Deployment
        name: test-comp-1-1
        uid: 3dedd956-d30a-4776-b863-e5ee6716f7a5
      traits:
      - customStatusMsg: |
          No loadBalancer found, visiting by using 'vela port-forward test-health-app-1 --route'
        healthStatus: HEALTHY
        resource: ""
        type: ingress
    - componentName: test-comp-1-2
      diagnosis: 'Ready:1/1 '
      healthStatus: HEALTHY
      targetWorkload:
        apiVersion: apps/v1
        kind: Deployment
        name: test-comp-1-2
        uid: 58d46d50-c116-49a3-b888-ba8d26fd8089
  - appName: test-health-app-2
    components:
    - componentName: test-comp-2-1
      diagnosis: 'Ready:1/1 '
      healthStatus: HEALTHY
      targetWorkload:
        apiVersion: apps/v1
        kind: Deployment
        name: test-comp-2-1
        uid: a6b691f0-fb83-4164-818a-39c76db61381
      traits:
      - customStatusMsg: |
          No loadBalancer found, visiting by using 'vela port-forward test-health-app-2 --route'
        healthStatus: HEALTHY
        resource: ""
        type: ingress
    - componentName: test-comp-2-2
      diagnosis: 'Ready:1/1 '
      healthStatus: HEALTHY
      targetWorkload:
        apiVersion: apps/v1
        kind: Deployment
        name: test-comp-2-2
        uid: 7f15a638-8a0e-471c-8aaa-0ea08202a174
  scopeHealthCondition:
    healthStatus: HEALTHY
    healthyWorkloads: 4
    total: 4

```

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go (hunk 1 above)
Comment (captainroy-hy) [reply pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go:136]: solved.

File: apis/core.oam.dev/v1alpha2/core_scope_types.go (hunk 2 above)
Comment (captainroy-hy) [reply apis/core.oam.dev/v1alpha2/core_scope_types.go]: fixed

File: apis/core.oam.dev/v1alpha2/core_scope_types.go [hunk 4]
```diff
@@ -47,8 +47,24 @@ type HealthScopeSpec struct {
 	// ProbeInterval is the amount of time in seconds between probing tries.
 	ProbeInterval *int32 `json:"probe-interval,omitempty"`
 
+	// AppRefs records references of applications' components
+	AppRefs AppReferences `json:"appReferences,omitempty"`
```
Comment (zzxwill) [inline apis/core.oam.dev/v1alpha2/core_scope_types.go:51]:
```suggestion
	AppRefs ComponentsReferences `json:"appReferences,omitempty"`
```
Seems you are referencing to components.

File: pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go (hunk 3 above)
Comment (captainroy-hy) [reply pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go:146]: fixed.

File: apis/core.oam.dev/v1alpha2/core_scope_types.go [hunk 5]
```diff
@@ -47,8 +47,24 @@ type HealthScopeSpec struct {
...
+	// +deprecated
+	WorkloadReferences []corev1.ObjectReference `json:"workloadRefs,omitempty"`
+}
+
+// AppReferences records references of applications' components
+type AppReferences map[string]CompReferences
+
+// CompReferences records references of components' resources
+type CompReferences map[string]ComponentReference
+
```
Comment (zzxwill) [inline apis/core.oam.dev/v1alpha2/core_scope_types.go:63]: Why a new type `CompReferences` is added?

File: apis/core.oam.dev/v1alpha2/core_scope_types.go [hunk 6]
```diff
@@ -47,8 +47,24 @@ type HealthScopeSpec struct {
...
+// ComponentReference records references of resources of a component
+type ComponentReference struct {
+	Workload corev1.ObjectReference   `json:"workload,omitempty"`
+	Traits   []corev1.ObjectReference `json:"traits,omitempty"`
```
Comment (zzxwill) [inline apis/core.oam.dev/v1alpha2/core_scope_types.go:67]: Can a healthscope check the status of more than one trait?

File: apis/core.oam.dev/v1alpha2/core_scope_types.go (hunk 5 above)
Comment (captainroy-hy) [reply apis/core.oam.dev/v1alpha2/core_scope_types.go:63]: CRD map type must use a named type as value type.

File: apis/core.oam.dev/v1alpha2/core_scope_types.go (hunk 6 above)
Comment (captainroy-hy) [reply apis/core.oam.dev/v1alpha2/core_scope_types.go:67]: Yes, actually, it just how current app controller CUE-based health checking mechanism works.

Comment (captainroy-hy) [reply apis/core.oam.dev/v1alpha2/core_scope_types.go:67]: BTW, these traits references are not used now. CUE-based health checking has its own way to get traits references.
This field will be used for non-CUE based health checking in HealthScope controller in the future.

**Comment by wonderflow** [comment]
I think it's better to align the format of the spec and status:

```
apiVersion: core.oam.dev/v1alpha2
kind: HealthScope
metadata:
...
spec:
  appReferences: # !! Pls help to check this field. It's used to record references of applications' components.
  - appName: test-health-app-1:
     components:
     - componentName: test-comp-1-1
        traits:
        - apiVersion: v1
          kind: Service
          name: test-comp-1-1
        - apiVersion: networking.k8s.io/v1beta1
          kind: Ingress
          name: test-comp-1-1
        workload:
          apiVersion: apps/v1
          kind: Deployment
          name: test-comp-1-1

  probe-interval: 1
  probe-timeout: 5
```

**Review by wonderflow — COMMENTED** [review]
the logic LGTM, please write tests

**Review by wonderflow — APPROVED** [review]
good job

**Pull request merged**